### PR TITLE
AppIntroViewPager: Fix slide policy handling when sliding the view pager

### PR DESCRIPTION
**Pull request by maxee**
Due to lack of overriding `onInterceptTouchEvent` and not overriding `onTouchEvent` correctly, sliding between slides is not getting disabled, even when the associated policy restricts sliding to the next/previous slide.

Somehow my solution provided in 2016/2018 has been exchanged for a wrong one. So I reimplemented the solution again (see #214 )

### CODE REVIEW

File: appintro/src/main/java/com/github/paolorotolo/appintro/AppIntroViewPager.kt [hunk 1]
```diff
@@ -97,45 +96,68 @@ class AppIntroViewPager(context: Context, attrs: AttributeSet) : ViewPager(conte
...
      * Util function to check if the user Illegaly request a swipe.
      * Also checks if the event happened not earlier than every 1000ms
      */
     private fun userIllegallyRequestNextPage(event: MotionEvent): Boolean {
-        if (Math.abs(event.x - currentTouchDownX) >= VALID_SWIPE_THRESHOLD_PX) {
-            if (System.currentTimeMillis() - illegallyRequestedNextPageLastCalled >=
-                ON_ILLEGALLY_REQUESTED_NEXT_PAGE_MAX_INTERVAL
-            ) {
+        if ((event.x - currentTouchDownX).absoluteValue >= VALID_SWIPE_THRESHOLD_PX_X
+                && (event.y - currentTouchDownY).absoluteValue <= VALID_SWIPE_THRESHOLD_PX_Y) {
```
Comment (AnuthaDev) [inline appintro/src/main/java/com/github/paolorotolo/appintro/AppIntroViewPager.kt]: Why exactly do we also need Y-Coordinate? Viewpager only supports Horizontal navigation.

Comment (maxee) [reply appintro/src/main/java/com/github/paolorotolo/appintro/AppIntroViewPager.kt]: @AnuthaDev It prevents detecting a false positive sliding attempt. Think of a scroll view within a slide. If one wants to scroll the scroll view vertically but also scrolls a bit horizontally (by accident), a sliding attempt would be assumed (if the user scrolled more than 25px horizontally during the intended vertical scroll).

File: appintro/src/main/java/com/github/paolorotolo/appintro/AppIntroViewPager.kt [hunk 2]
```diff
@@ -97,45 +96,68 @@ class AppIntroViewPager(context: Context, attrs: AttributeSet) : ViewPager(conte
         customScroller?.scrollDurationFactor = factor
     }
 
-    override fun performClick() = super.performClick()
+    override fun onInterceptTouchEvent(event: MotionEvent?): Boolean
+    {
+        if (!handleTouchEvent(event)) {
+            return false
```
Comment (cortinico) [inline appintro/src/main/java/com/github/paolorotolo/appintro/AppIntroViewPager.kt]: You need to return `true` here.

File: appintro/src/main/java/com/github/paolorotolo/appintro/AppIntroViewPager.kt [hunk 3]
```diff
@@ -97,45 +96,68 @@ class AppIntroViewPager(context: Context, attrs: AttributeSet) : ViewPager(conte
...
+        }
+
+        // Calling super will allow the slider to "work" left and right.
+        return super.onInterceptTouchEvent(event)
+    }
 
     override fun onTouchEvent(event: MotionEvent?): Boolean {
+        if (!handleTouchEvent(event)) {
+            return false
```
Comment (cortinico) [inline appintro/src/main/java/com/github/paolorotolo/appintro/AppIntroViewPager.kt]: You need to return `true` here.

File: appintro/src/main/java/com/github/paolorotolo/appintro/AppIntroViewPager.kt [hunk 4]
```diff
@@ -97,45 +96,68 @@ class AppIntroViewPager(context: Context, attrs: AttributeSet) : ViewPager(conte
...
+    private fun handleTouchEvent(event: MotionEvent?) : Boolean {
         // If paging is disabled we should ignore any viewpager touch
         // (also, not display any error message)
         if (!isFullPagingEnabled) {
             return false
         }
 
-        val canRequestNextPage = onNextPageRequestedListener?.onCanRequestNextPage() ?: true
-        when (event?.action) {
-            MotionEvent.ACTION_UP -> performClick()
```
Comment (cortinico) [inline appintro/src/main/java/com/github/paolorotolo/appintro/AppIntroViewPager.kt]: Please keep this line for accessibility purposes.

File: appintro/src/main/java/com/github/paolorotolo/appintro/AppIntroViewPager.kt [hunk 5]
```diff
@@ -97,45 +96,68 @@ class AppIntroViewPager(context: Context, attrs: AttributeSet) : ViewPager(conte
...
     override fun onTouchEvent(event: MotionEvent?): Boolean {
+        if (!handleTouchEvent(event)) {
+            return false
+        }
+
+        // Calling super will allow the slider to "work" left and right.
+        return super.onTouchEvent(event)
+    }
+
+    private fun handleTouchEvent(event: MotionEvent?) : Boolean {
```
Comment (lucalves) [inline appintro/src/main/java/com/github/paolorotolo/appintro/AppIntroViewPager.kt]: I would recommend a comment specifying this function.

**Comment by cortinico** [comment]
@maxee Any updates?

File: appintro/src/main/java/com/github/paolorotolo/appintro/AppIntroViewPager.kt (hunk 5 above)
Comment (maxee) [reply appintro/src/main/java/com/github/paolorotolo/appintro/AppIntroViewPager.kt]: Done

**Comment by maxee** [comment]
@cortinico I added another commit, fixing most of the comments, i guess.

**Comment by cortinico** [comment]
> @cortinico I added another commit, fixing most of the comments, i guess.

Great :) How about the other comments? If you don't have time to finish it, I can also pick up this branch from where you left it.

**Comment by cortinico** [comment]
I'm taking over the branch and merging it as this PR was opened long ago and it makes sense to have this fix in master :)

**Pull request merged**